doc/README.md

```diff
@@ -0,0 +1,421 @@
+      "length": 9455314,
+      "uncompressedLength": 9855314
+    },
+    "4f34352be7c1f4d1597d27f5824b42bad2222cf029cc888e9d7318d5e8bc2ad5": {
+      "id": "a79d2b9e7afa051782cff6424ede1337c8ce05c34d9c2e565e2ed47e7ae80ea4",
+      "length": 8430346,
+      "uncompressedLength": 9430346
+    },
+    "16759b52e8ef8c3b080f0a8cdb8ac21474cb41070f272a02e28fd1163762336b": {
+      "id": "31aa8e88c8b73df565fd8bbcadfe7e2e79303eb5c0223253353d3003273e3140",
+      "length": 26549052,
+      "uncompressedLength": 28549052
+    },
+    "c11c3cf1c326375b177e768de908e6e423c3d8866715c8ec7159156b6ea82497": {
+      "id": "5421895dbead0ba83f0993f17b5f72f6c6618c0e9ea849c6c53ac240d1802d0b",
+      "length": 523496,
+      "uncompressedLength": 645314
+    }
+  }
+}
+```
+
+The `chunkIds` and `iconChunkIds` fields contain a list with plain text SHA-256 hashes
+which can be found in the main `blobs` dictionary.
+This contains a mapping from plain text SHA-256 hashes to storage IDs and size information.
+
+At the beginning of most operations, we download all available snapshots
+to get information about all blobs that should be available in the repository.
+We may additionally retrieve a list of all blobs directly from the repository
+to ensure they are actually (still) present.
+
+Snapshot files start with the SHA-256 hash of the content of the file
+and use a `.snapshot` extension.
+
+## Cryptography
+
+This section is based on and thus very similar to encryption of
+[files backup](../storage/doc/design.md).
+
+### Main Key
+
+Seedvault already uses [BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki)
+to give users a mnemonic recovery code and for generating deterministic keys.
+The derived key has 512 bits
+and Seedvault used to use the first 256 bits as an AES key to encrypt app data.
+Unfortunately, this key's usage is limited by Android's keystore to encryption and decryption.
+Therefore, the second 256 bits are be imported into Android's keystore for use with `HMAC-SHA256`,
+so that this key can act as a main key we can deterministically derive additional keys from
+by using HKDF ([RFC5869](https://tools.ietf.org/html/rfc5869)).
+These second 256 bits must not be used for any other purpose in the future.
+We use them for a main key to avoid users having to handle yet another secret.
+
+For deriving keys, we are only using the HKDF's second 'expand' step,
+because the Android Keystore does not give us access
+to the key's byte representation (required for first 'extract' step) after importing it.
+This should be fine as the input key material is already a cryptographically strong key
+(see section 3.3 of RFC 5869 above).
+
+### Key derivation overview
+
+The original entropy comes from a BIP39 seed (12 words = 128 bit size)
+obtained from Java's `SecureRandom`.
+A PBKDF SHA512 based derivation defined in BIP39 turns this into a 512 bit seed key.
+
+The derived seed key (512 bit size) gets split into two parts:
```

Comment:
[Clarity] Nit: Some of this section seems to duplicate the initial Main Key section. Consider adding a phrase like "As referenced earlier" to ensure that the reader easily understands that this section is discussing the same thing, rather than something new or subtly different.